Robust partition names

Spacerat opened the pull request:
w.r.t: #335 

The primary purpose of this diff is to ensure that partition names such as `2018-10-10` get parsed as dates instead of integer literals when reading hive-partitioned parquet files.

More specifically:
- Partition names containing any of `+boxjeBOXJE_'"` are no longer parsed as string literals. I would be extremely surprised if anyone was relying on scientific notation, imaginary numbers, arithmetic, etc. in partition names
- If any two partition names map to the same value e.g. `"0.7"` and `".7"` or `"may 9 2018"` and `"2018-05-09"`, an informative ValueError is thrown
- If any the number of types encountered in the partition values after transformation by `val_to_num()` is greater than 1, a warning is displayed.

(There's also a new test for filtering by dates.)

As discussed in the issue, a future feature could be added to enable/disable this inference entirely, either across all partitions or on a per-name basis. However, I think it would be good to at least get this in for now.

martindurant commented:
Just in: pandas 0.23.0 recognises `'now'` as a time.
I don't think that is what causes the test failures, though - seems important in `fastparquet/test/test_partition_filters_specialstrings.py::test_frame_write_read_verify`.

Spacerat commented:
Ah, yep, I'd neglected to run the test_output tests locally, so I've fixed some mistakes for them. You're right about test_partition_filters_spectialstrings though; that's definitely very broken. I'll try to fix everything locally then push again

Spacerat commented:
@martindurant you were right about `'now'` being the issue there. I added it as another special case to to `utils.py` and fixed the other issues and now all the tests pass.

I also came up with what I think is a reasonable rationale for thinking about `val_to_num`: if fastparquet will never write a certain literal, we probably shouldn't parse it. This rules out set/list/dict literals, bin/hex/oct/imaginary numbers and, of course arithmetic expressions; as a result, I decided to stop using `literal_eval` since it was getting kind of silly to just keep adding restricted characters as I found the need for them. Instead I just use `float()` and `int()`

However, I brought back support for scientific notation in literals, since if you do partition on extremely large or small numbers, the directory names will actually be written out using scientific notation. I've also added some tests for that case.

martindurant commented:
This looks good with thorough testing, thank you.

Spacerat commented:
Thanks!
